StackExchange.Redis/StackExchange/Redis/ChannelMessageQueue.cs

```diff
@@ -0,0 +1,198 @@
+        }
+        /// <summary>
+        /// Create a message loop that processes messages sequentially
+        /// </summary>
+        public void OnMessage(Action<RedisChannel, RedisValue> handler)
+        {
+            AssertOnMessage(handler);
+            ThreadPool.QueueUserWorkItem(
+                state => ((ChannelMessageQueue)state).OnMessageSyncImpl(), this);
+        }
+
+        private async void OnMessageSyncImpl()
+        {
+            var handler = (Action<RedisChannel, RedisValue>)_onMessageHandler;
+            while (!IsCompleted)
+            {
+                ChannelMessage next;
+                try { if(!TryRead(out next)) next = await ReadAsync(); }
+                catch (ChannelClosedException) { break; } // expected
+                catch (Exception ex)
+                {
+                    _parent.multiplexer?.OnInternalError(ex);
+                    break;
+                }
+
+                try { handler.Invoke(next.Channel, next.Value); }
+                catch { } // matches MessageCompletable
+            }
+        }
+
+        /// <summary>
+        /// Create a message loop that processes messages sequentially
+        /// </summary>
+        public void OnMessage(Func<RedisChannel, RedisValue, Task> handler)
+        {
+            AssertOnMessage(handler);
+            ThreadPool.QueueUserWorkItem(
+                state => ((ChannelMessageQueue)state).OnMessageAsyncImpl(), this);
+        }
+
+        private async void OnMessageAsyncImpl()
+        {
+            var handler = (Func<RedisChannel, RedisValue, Task>)_onMessageHandler;
+            while (!IsCompleted)
+            {
+                ChannelMessage next;
+                try { if (!TryRead(out next)) next = await ReadAsync(); }
+                catch (ChannelClosedException) { break; } // expected
+                catch (Exception ex)
+                {
+                    _parent.multiplexer?.OnInternalError(ex);
+                    break;
+                }
+
+                try
+                {
+                    var task = handler.Invoke(next.Channel, next.Value);
+                    if (task != null) await task;
+                }
+                catch { } // matches MessageCompletable
+            }
+        }
+        internal void UnsubscribeImpl(Exception error = null, CommandFlags flags = CommandFlags.None)
+        {
+            var parent = _parent;
+            if (parent != null)
+            {
+                _parent.UnsubscribeAsync(_redisChannel, HandleMessage, flags);
+                _parent = null;
+                _channel.Writer.TryComplete(error);
+            }
+        }
+        internal async Task UnsubscribeAsyncImpl(Exception error = null, CommandFlags flags = CommandFlags.None)
+        {
+            var parent = _parent;
+            if (parent != null)
+            {
+                await _parent.UnsubscribeAsync(_redisChannel, HandleMessage, flags);
```

Comment:
There's a race in these - if someone called it concurrently, `_parent` could be a null ref - could avoid by using `parent.UnsubscribeAsync` here (similar in the other equivalents).

Reply:
was meant to be that, thanks; good eyes